javalib/src/main/scala/java/util/Collections.scala

```diff
@@ -0,0 +1,1194 @@
+          protected def inner: Set[Map.Entry[K, V]] = innerSet
+
+          override def iterator(): Iterator[Map.Entry[K, V]] = {
+            val innerIterator = super.iterator()
+            new WrappedIterator[Map.Entry[K, V], Iterator[Map.Entry[K, V]]] {
+              protected def inner: Iterator[Map.Entry[K, V]] = innerIterator
+
+              override def next(): Map.Entry[K, V] = {
+                val nextEntry = super.next()
+                new Map.Entry[K, V] {
+                  def getKey(): K =
+                    nextEntry.getKey()
+
+                  def getValue(): V =
+                    nextEntry.getValue()
+
+                  def setValue(value: V): V = {
+                    CheckedCollections.checkClass(value, valueClazz)
+                    nextEntry.setValue(value)
+                  }
+
+                  override def equals(o: Any): Boolean =
+                    nextEntry.equals(o)
+
+                  override def hashCode(): Int =
+                    nextEntry.hashCode()
+                }
+              }
+            }
+          }
+        }
+      }
+
+      protected final def checkKeyAndValue(key: K, value: V) = {
+        CheckedCollections.checkClass(key, keyClazz)
+        CheckedCollections.checkClass(value, valueClazz)
+      }
+    }
+
+    private[Collections] trait CheckedSortedMap[K, V]
+        extends CheckedMap[K, V, SortedMap[K, V]] with WrappedSortedMap[K, V] {
+      override def subMap(fromKey: K, toKey: K): SortedMap[K, V] =
+        checkedSortedMap(super.subMap(fromKey, toKey), keyClazz, valueClazz)
+
+      override def headMap(toKey: K): SortedMap[K, V] =
+        checkedSortedMap(super.headMap(toKey), keyClazz, valueClazz)
+
+      override def tailMap(fromKey: K): SortedMap[K, V] =
+        checkedSortedMap(super.tailMap(fromKey), keyClazz, valueClazz)
+    }
+
+    private class CheckedListIterator[E](protected val inner: ListIterator[E],
+        protected val elemClazz: Class[E]) extends WrappedListIterator[E] {
+      override def set(e: E): Unit = {
+        checkElem(e)
+        super.set(e)
+      }
+
+      override def add(e: E): Unit = {
+        checkElem(e)
+        super.add(e)
+      }
+
+      private def checkElem(elem: E): Unit = {
+        checkClass(elem, elemClazz)
+      }
+    }
+  }
+
+  private class EmptyIterator extends Iterator[Any] {
+    def hasNext: Boolean = false
+
+    def next(): Any =
+      throw new NoSuchElementException
+
+    def remove(): Unit =
+      throw new IllegalStateException
+  }
+
+  private class EmptyListIterator extends EmptyIterator with ListIterator[Any] {
+    def hasPrevious: Boolean = false
+
+    def previous(): Any =
+      throw new NoSuchElementException
+
+    def nextIndex(): Int = 0
+
+    def previousIndex(): Int = -1
+
+    def set(e: Any): Unit =
+      throw new IllegalStateException
+
+    def add(e: Any): Unit =
+      throw new UnsupportedOperationException
```

Comment:
They where defined like that in the doc and the JVM agrees:
```
scala> val e = java.util.Collections.emptyListIterator[Int]
e: java.util.ListIterator[Int] = java.util.Collections$EmptyListIterator@6f2b93de

scala> e.add(1)
java.lang.UnsupportedOperationException
  at java.util.Collections$EmptyListIterator.add(Collections.java:4240)
  ... 43 elided
```

```
scala> val e = java.util.Collections.emptyListIterator[Int]
e: java.util.ListIterator[Int] = java.util.Collections$EmptyListIterator@56fe88e4

scala> e.set(1)
java.lang.IllegalStateException
  at java.util.Collections$EmptyListIterator.set(Collections.java:4239)
  ... 43 elided
```
The `set` is in an illegal state because it requires a call to `next` or `previous` first, while the add can insert at any time.